python/pyspark/sql/connect/functions.py

```diff
@@ -2350,8 +2356,21 @@ def unwrap_udt(col: "ColumnOrName") -> Column:
 unwrap_udt.__doc__ = pysparkfuncs.unwrap_udt.__doc__
 
 
-def udf(*args: Any, **kwargs: Any) -> None:
-    raise NotImplementedError("udf() is not implemented.")
+def udf(
+    f: Optional[Union[Callable[..., Any], "DataTypeOrString"]] = None,
+    returnType: "DataTypeOrString" = StringType(),
+) -> Union["UserDefinedFunctionLike", Callable[[Callable[..., Any]], "UserDefinedFunctionLike"]]:
+    if f is None or isinstance(f, (str, DataType)):
+        # If DataType has been passed as a positional argument
+        # for decorator use it as a returnType
+        return_type = f or returnType
+        return functools.partial(
+            _create_udf, returnType=return_type, evalType=100  # PythonEvalType.SQL_BATCHED_UDF
```

Comment:
I see that this class is defined in rdd.py. @HyukjinKwon can we create a PR to move this constant definition out of this file into a more generic place?

Reply:
That's actually also used in Python RDD, and Python UDF is based on the top of RDD so it's not that bad :-).

Xinrong, you can import `PythonEvalType.SQL_BATCHED_UDF` within this function to avoid cyclic import

Reply:
Sounds good, `PythonEvalType.SQL_BATCHED_UDF` is imported inside the function.